src/tests/suite/daos_rebuild.c

```diff
@@ -348,6 +372,157 @@ rebuild_pool_destroy(test_arg_t *arg)
+	if (arg->myrank == 0) {
+		rc = daos_pool_connect(arg->pool.pool_uuid, arg->group,
+				       &arg->pool.svc, DAOS_PC_RW,
+				       &arg->pool.poh, &arg->pool.pool_info,
+				       NULL /* ev */);
+		if (rc)
+			print_message("daos_pool_connect failed, rc: %d\n", rc);
+
+		print_message("pool connect "DF_UUIDF"\n",
+			       DP_UUID(arg->pool.pool_uuid));
+	}
+	MPI_Barrier(MPI_COMM_WORLD);
+	if (arg->multi_rank)
+		MPI_Bcast(&rc, 1, MPI_INT, 0, MPI_COMM_WORLD);
+	if (rc)
+		return rc;
+
+	/** broadcast pool info */
+	if (arg->multi_rank) {
+		MPI_Bcast(&arg->pool.pool_info, sizeof(arg->pool.pool_info),
+			  MPI_CHAR, 0, MPI_COMM_WORLD);
+		handle_share(&arg->pool.poh, HANDLE_POOL, arg->myrank,
+			     arg->pool.poh, 0);
+	}
+
+	/** open container */
+	MPI_Barrier(MPI_COMM_WORLD);
+	if (arg->myrank == 0) {
+		rc = daos_cont_open(arg->pool.poh, arg->co_uuid, DAOS_COO_RW,
+				    &arg->coh, &arg->co_info, NULL);
+		if (rc)
+			print_message("daos_cont_open failed, rc: %d\n", rc);
+
+		print_message("container open "DF_UUIDF"\n",
+			       DP_UUID(arg->co_uuid));
+	}
+	MPI_Barrier(MPI_COMM_WORLD);
+	if (arg->multi_rank)
+		MPI_Bcast(&rc, 1, MPI_INT, 0, MPI_COMM_WORLD);
+	if (rc)
+		return rc;
+
+	/** broadcast container info */
+	if (arg->multi_rank) {
+		MPI_Bcast(&rc, 1, MPI_INT, 0, MPI_COMM_WORLD);
+		handle_share(&arg->coh, HANDLE_CO, arg->myrank, arg->pool.poh,
+			     0);
+	}
+
+	return 0;
+}
+
+static int
+rebuild_pool_disconnect_internal(void *data)
+{
+	test_arg_t	*arg = data;
+	int		rc = 0;
+	int		rc_reduce = 0;
+
+	/* Close cont and disconnect pool */
+	rc = daos_cont_close(arg->coh, NULL);
+	if (arg->multi_rank) {
+		MPI_Allreduce(&rc, &rc_reduce, 1, MPI_INT, MPI_MIN,
+			      MPI_COMM_WORLD);
+		rc = rc_reduce;
+	}
+	print_message("container close "DF_UUIDF"\n",
+		      DP_UUID(arg->co_uuid));
+	if (rc) {
+		print_message("failed to close container "DF_UUIDF
+			      ": %d\n", DP_UUID(arg->co_uuid), rc);
+		return rc;
+	}
+
+	arg->coh = DAOS_HDL_INVAL;
+	rc = daos_pool_disconnect(arg->pool.poh, NULL /* ev */);
+	if (rc)
+		print_message("failed to disconnect pool "DF_UUIDF
+			      ": %d\n", DP_UUID(arg->pool.pool_uuid), rc);
+
+	print_message("pool disconnect "DF_UUIDF"\n",
+		      DP_UUID(arg->pool.pool_uuid));
+
+	arg->pool.poh = DAOS_HDL_INVAL;
+	MPI_Barrier(MPI_COMM_WORLD);
+	return rc;
+}
+
+void
+reintegrate_single_pool_target(test_arg_t *arg, d_rank_t failed_rank,
+			       int failed_tgt)
+{
+	/* XXX: Disconnecting and reconnecting is necessary for the time being
+	 * while reintegration only supports "offline" mode and without
+	 * incremental reintegration. Disconnecting from the pool allows the
+	 * containers to be deleted before reintegration occurs
+	 *
+	 * Once incremental reintegration support is added, this should be
+	 * removed
+	 */
+	rebuild_pool_disconnect_internal(arg);
+	rebuild_targets(&arg, 1, &failed_rank, &failed_tgt, 1, false, false);
+	rebuild_pool_connect_internal(arg);
```

Comment:
I am not sure if you want to add reintegration to so many rebuild test. If CI can pass, I am fine with it.
But it seems there are still lot failures in the last run. So let's confine this maybe only to 1 or 2 test cases for the moment.  Please check.
https://build.hpdd.intel.com/job/daos-stack/job/daos/view/change-requests/job/PR-1802/24/testReport/junit/(root)/DaosCoreTestRebuild/Test___Functional_Hardware_Medium___01___daos_test_daos_core_test_rebuild_py_DaosCoreTestRebuild_test_rebuild_test_r_0_10_num_clients_num_replicas_hosts_0_1_eba7/

Reply:
> ADD_IN would only be set after all reintegration is finished, i.e. it is set through the reintegrate process,
I don't believe POOL_ADD_IN is actually sent as an RPC anymore - but the opcode declared here is re-used by the ds_pool_tgt_add_in() function that rebuild uses, which match the other exclude and exclude_out calls. I could add a new enum for this... but it seems messy. This might still be useful for tests in the future (i.e. using the same mechanic rebuild used prior to this patch of forcibly setting targets back to UPIN when the data doesn't matter). 

> Liang: what race do you mean here?
We dug through this code on the Tuesday call last week and decided there was not a race - things are being called on the correct xstream. 

> I am not sure if you want to add reintegration to so many rebuild test.
I believe I found the fix for this (checking if pool was destroyed before connecting to check if rebuild is complete). If it still doesn't pass after this I'll pull these out and go back to POOL_ADD_IN RPCs for many of those tests.